# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Desiree Blake https://github.com/dblake26/json-sentiment 

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [21]:

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package                  Version
------------------------ -----------
annotated-types          0.6.0
anyio                    3.6.2
argon2-cffi              21.3.0
argon2-cffi-bindings     21.2.0
arrow                    1.2.3
asttokens                2.2.1
async-lru                2.0.2
attrs                    23.1.0
Babel                    2.12.1
backcall                 0.2.0
beautifulsoup4           4.12.2
black                    23.3.0
bleach                   6.0.0
blis                     0.7.11
build                    0.10.0
catalogue                2.0.10
certifi                  2024.2.2
cffi                     1.15.1
charset-normalizer       3.3.2
click                    8.1.7
cloudpathlib             0.16.0
colorama                 0.4.6
comm                     0.1.3
confection               0.1.4
contourpy                1.2.0
cycler                   0.12.1
cymem                    2.0.8
debugpy                  1.6.7
decorator               

Question 1

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

Note: The above instructions are not longer working so we will use https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd // author/poem instead

In [22]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import requests
import json

AUTHOR='Edgar Allan Poe'
POEM = 'A Dream Within A Dream'

#only certain poets and titles are available
#to see the available poets, go to (in a web browser)
# https://poetrydb.org/author
#To see which poems that author has available, go to 
# https://poetrydb.org/author/AUTHOR NAME
# e.g.: https://poetrydb.org/author/Edgar Allan Poe
#The spaces will get handled by your web browser

# A cool pythonism (introduced in Python 3): f strings
# https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings
URL = f'https://poetrydb.org/author,title/{AUTHOR};{POEM}'
result = json.loads(requests.get(URL).text)

# write the result to a JSON file
with open('poem.json', 'w') as file:
    json.dump(result, file)
    
# Print the poem to get results
poem = '\n'.join(result[0]['lines']) 
print(poem)


Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?


Question 2

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

Note: The above instructions are not longer working so we will use https://gist.github.com/neloe/e38e4d3283418e096aac80fa48bb66bd // author/poem instead

In [23]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Read in the contents of the JSON file
with open('poem.json', 'r') as file:
    result = json.load(file)

# Extract the poem's text 
poem_lines = result[0]['lines']
poem_text = '\n'.join(poem_lines)

# Print the lyrics of the poem
print("Poem Lyrics:\n", poem_text)

# Perform sentiment analysis on the poem
doc = nlp(poem_text)
polarity_score = doc._.polarity

# Print the polarity score of the sentiment analysis
print("\nPolarity Score:", polarity_score)

# With a polarity score of  0.055555555555555546, I would say this is fairly neutral but leaning more towards a positive sentiment. 

Poem Lyrics:
 Take this kiss upon the brow!
And, in parting from you now,
Thus much let me avow--
You are not wrong, who deem
That my days have been a dream:
Yet if hope has flown away
In a night, or in a day,
In a vision or in none,
Is it therefore the less _gone_?
_All_ that we see or seem
Is but a dream within a dream.

I stand amid the roar
Of a surf-tormented shore,
And I hold within my hand
Grains of the golden sand--
How few! yet how they creep
Through my fingers to the deep
While I weep--while I weep!
O God! can I not grasp
Them with a tighter clasp?
O God! can I not save
_One_ from the pitiless wave?
Is _all_ that we see or seem
But a dream within a dream?

Polarity Score: 0.055555555555555546


Question 3

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.


In [24]:
import requests
import json

def write_song_from_api_to_json(artist, song, filename):
    url = f"https://api.lyrics.ovh/v1/{artist}/{song}"
    response = requests.get(url)
    if response.status_code == 200:
        data = response.json()
        with open(filename, 'w') as file:
            json.dump(data, file)
            print(f"Lyrics for {song} by {artist} written to {filename}")
    else:
        print(f"Failed to retrieve lyrics for {song} by {artist}")

# Test the function with four songs
write_song_from_api_to_json("Stevie Wonder", "For Once in My Life", "for_once_in_my_life_lyrics.json")
write_song_from_api_to_json("SZA", "Snooze", "snooze_lyrics.json")
write_song_from_api_to_json("Future", "LOVE YOU BETTER", "LOVE_YOU_BETTER_lyrics.json")
write_song_from_api_to_json("Snoh Aalegra", "Whoa", "whoa_lyrics.json")



Lyrics for For Once in My Life by Stevie Wonder written to for_once_in_my_life_lyrics.json
Lyrics for Snooze by SZA written to snooze_lyrics.json
Lyrics for LOVE YOU BETTER by Future written to LOVE_YOU_BETTER_lyrics.json
Lyrics for Whoa by Snoh Aalegra written to whoa_lyrics.json


Question 4

4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.


In [25]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

def calculate_sentiment_polarity(filename):
    with open(filename, 'r') as file:
        data = json.load(file)
        poem = data['lyrics']

    nlp = spacy.load('en_core_web_sm')
    nlp.add_pipe('spacytextblob')
    doc = nlp(poem)
    return doc._.polarity

# Calculate and print the polarity scores for the three files
filenames = filenames = ["for_once_in_my_life_lyrics.json", "snooze_lyrics.json", "LOVE_YOU_BETTER_lyrics.json", "whoa_lyrics.json"]

for filename in filenames:
    polarity_score = calculate_sentiment_polarity(filename)
    print(f"{filename}: Polarity Score: {polarity_score}")
    
# 

for_once_in_my_life_lyrics.json: Polarity Score: 0.24259259259259258
snooze_lyrics.json: Polarity Score: 0.14413265306122441
LOVE_YOU_BETTER_lyrics.json: Polarity Score: 0.44062499999999993
whoa_lyrics.json: Polarity Score: 0.17142857142857143


Question 4 Reflections

| Song Title | Polarity Score | Does Report Polarity Match My Understanding of the Song? | 
|-----------------|-----------------|-----------------|
| For Once in My Life | 0.24259259259259258 | The score suggests it is slightly positive so no. I agree but I expected it to be the most positive in this list.| 
| Snooze | 0.14413265306122441 | I think this song is fair being neutral but i did expect it to be more positive with lyrics like "I can lose when i'm with you". |
| LOVE YOU BETTER | 0.44062499999999993 | Considering this song is about someone falling out of love, I did not expect it to be the most positive of the list. I do not agree but see how this can happen because the song does have the words "love" repeated often. | 
| Whoa | 0.17142857142857143 | This one does match my understanding with being fairly positive but mostly neutral. |
